boards: m5stack: cores3: Add and Enable AXP192/AXP2101 LED driver

soburi opened the pull request:
Enabling AXP2101 LED Controller on M5Stack CoreS3.

- Add AXP192/AXP2101 LED Controller driver.
- Some updating about AXP192 MFD driver.
  - Define bus for describe internal dependency
  - Enhance it to support for AXP2101

I verified it only M5Stack CoreS3 SE/AXP2101.

soburi committed:
drivers: led: Add driver for AXP192/2101 LED control function

Add a driver to support AXP192/2101's LED control function.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
boards: m5stack: cores3: Add configuration for AXP2101

Adding configuration for enabling AXP2101.
Currently, it supports only LED features.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
tests: driver: build_all: led: add AXP192/2101 to tests

Add AXP192/2101 LED controller function driver to build_all test.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
dts: bindings: Define a bus for the axp192/axp2101.

Define a bus to show the dependencies of each function
on the AXP192/AXP2101 MFD.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

gmarull commented on dts/bindings/gpio/x-powers,axp192-gpio.yaml:
```diff
@@ -18,6 +18,8 @@ compatible: "x-powers,axp192-gpio"
 
 include: gpio-controller.yaml
 
+on-bus: axp192
```
can you please explain why this is a bus?

soburi replied:
I found this way in the max31790 implementation.
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/mfd/maxim%2Cmax31790.yaml#L10

I adopted it because I think it is a smart way to describe the MFD device and each function's dependencies.
But if this is not a good idea, I'll revert it.

simonguinot commented:
Hi @soburi. This PR is on my radar and I'll review it soon !

soburi commented:
@simonguinot 

Sorry for the wait.
I change this PR to available for review. Could you check it out when you have time?

simonguinot commented on dts/bindings/led/x-powers,axp192-led.yaml:
```diff
@@ -0,0 +1,21 @@
+      - "by-reg"
+    description:
+      Select the LED control method.
+      If you select "by-reg", you can control it from software.
```
Because so far "by-reg" is the only possible selection here :)

soburi replied:
ditto.

simonguinot reviewed (commented):
Hi @soburi,

Thanks for this driver.

I don't have much to say. Please see my comments and questions below.

soburi commented on drivers/led/led_axp192.c:
```diff
@@ -0,0 +1,139 @@
+	case CHGLED_CTRL_TYPE_B:
+	case CHGLED_CTRL_BY_REG:
+		return i2c_reg_update_byte_dt(&config->i2c, config->addr, config->mode_mask,
+					      config->mode << config->mode_offset);
```
At the moment there is no API that can do this, and settings other than by-reg are not compatible with this API, so I think this is fine for now.

simonguinot replied:
OK !

soburi commented on drivers/led/Kconfig.axp192:
```diff
@@ -0,0 +1,12 @@
+# Copyright (c) 2024 TOKITA Hiroshi
+# SPDX-License-Identifier: Apache-2.0
+
+config LED_AXP192_AXP2101
```
This follows the existing rule.

https://github.com/zephyrproject-rtos/zephyr/blob/d453fb94075c5d107f521f8a1ebaf85f3ead32d7/drivers/mfd/Kconfig.axp192#L4-L10

Although AXP192 and AXP2101 can be partially shared, there is no uniformity in the model numbers, so we have done it this way.

simonguinot replied:
OK !

However note that the names of Kconfig options don't relate with hardware compatibility but to driver support. If the driver can be called `led_axp192`, then so can the Kconfig option :)

But it is good to have the Kconfig naming consistent between LED and MFD.

soburi commented on dts/bindings/led/x-powers,axp192-led.yaml:
```diff
@@ -0,0 +1,21 @@
+  x-powers,mode:
+    type: string
+    enum:
+      - "by-reg"
```
Good point.
I added 'by-charge' for another configuration.

soburi commented on drivers/led/Kconfig.axp192:
```diff
@@ -0,0 +1,12 @@
+	select I2C
+	select MFD
+	help
+	  Enable the AXP192 LED driver.
```
I made the explanation a little more detailed.

soburi commented:
At this time, the Twister build is failing due to https://github.com/zephyrproject-rtos/zephyr/issues/88396.
I will re-run after the issue is fixed.

simonguinot commented on drivers/led/led_axp192.c:
```diff
@@ -27,9 +27,10 @@ LOG_MODULE_REGISTER(led_axp192, CONFIG_LED_LOG_LEVEL);
+#define CHGLED_CTRL_TYPE_A    0x0
+#define CHGLED_CTRL_TYPE_B    0x1
+#define CHGLED_CTRL_BY_REG    0x2
+#define CHGLED_CTRL_BY_CHARGE 0x3
```
I think you merged this changes (and those below) into the wrong commit.

soburi replied:
Nice catch!
I fixed and pushed it.
Thank you.

soburi commented:
A little modification is needed since #88125 merged.

soburi commented:
@simonguinot 

Added the following lines to coincide with the merge of #88125.

https://github.com/zephyrproject-rtos/zephyr/pull/82518/files#diff-dd1b1555daacb920efa045ef4b76714f223ed7d6823dc38502c2099aa3272bd9R26

> [nitpick] The error message in the blink function hardcodes valid delay values, which might become inconsistent if the constant definitions change. Consider referencing the defined constants or clearly documenting how these values relate to SLOW_BLINK_DELAY_ON/OFF and FAST_BLINK_DELAY_ON/OFF.

Addressed.

simonguinot reviewed (approved):
Thanks for this driver @soburi !
